src/Catalyst.Core.Modules.Keystore/KeyStoreService.cs

```diff
@@ -0,0 +1,830 @@
+        {
+            // TODO: Rename to GetIpfsPublicKeyAsync
+            string result = null;
+            var ekey = await Store.TryGetAsync(name, cancel).ConfigureAwait(false);
+            if (ekey != null)
+            {
+                UseEncryptedKey(ekey, key =>
+                {
+                    var kp = GetKeyPairFromPrivateKey(key);
+                    var spki = SubjectPublicKeyInfoFactory.CreateSubjectPublicKeyInfo(kp.Public).GetDerEncoded();
+
+                    // Add protobuf cruft.
+                    var publicKey = new PublicKey
+                    {
+                        Data = spki,
+                        Type = kp.Public switch
+                        {
+                            RsaKeyParameters _ => KeyType.Rsa,
+                            Ed25519PublicKeyParameters _ => KeyType.Ed25519,
+                            ECPublicKeyParameters _ => KeyType.Secp256K1,
+                            _ => throw new NotSupportedException(
+                                $"The key type {kp.Public.GetType().Name} is not supported.")
+                        }
+                    };
+
+                    using var ms = new MemoryStream();
+                    Serializer.Serialize(ms, publicKey);
+                    result = Convert.ToBase64String(ms.ToArray());
+                });
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public async Task<IKey> CreateAsync(string name,
+            string keyType,
+            int size,
+            CancellationToken cancel = default)
+        {
+            // Apply defaults.
+            if (string.IsNullOrWhiteSpace(keyType))
+            {
+                keyType = Options.DefaultKeyType;
+            }
+
+            if (size < 1)
+            {
+                size = Options.DefaultKeySize;
+            }
+
+            keyType = keyType.ToLowerInvariant();
+
+            // Create the key pair.
+            Log.DebugFormat("Creating {0} key named '{1}'", keyType, name);
+            IAsymmetricCipherKeyPairGenerator g;
+            switch (keyType)
+            {
+                case "rsa":
+                    g = GeneratorUtilities.GetKeyPairGenerator("RSA");
+                    g.Init(new RsaKeyGenerationParameters(
+                        BigInteger.ValueOf(0x10001), new SecureRandom(), size, 25));
+                    break;
+                case "ed25519":
+                    g = GeneratorUtilities.GetKeyPairGenerator("Ed25519");
+                    g.Init(new Ed25519KeyGenerationParameters(new SecureRandom()));
+                    break;
+                case "secp256k1":
+                    g = GeneratorUtilities.GetKeyPairGenerator("EC");
+                    g.Init(new ECKeyGenerationParameters(SecObjectIdentifiers.SecP256k1, new SecureRandom()));
+                    break;
+                default:
+                    throw new Exception($"Invalid key type '{keyType}'.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)